nemo/collections/asr/modules/conformer_encoder.py

```diff
@@ -384,6 +408,46 @@ def forward(self, audio_signal, length, cache_last_channel=None, cache_last_time
         max_audio_length = audio_signal.size(1)
 
         # Create the self-attention and padding masks
+        audio_signal, max_audio_length, pad_mask, att_mask, pos_emb, cache_last_channel_next = self._create_masks(
+            audio_signal, max_audio_length, length, cache_last_channel=cache_last_channel
+        )
+
+        for lth, layer in enumerate(self.layers):
+            audio_signal = layer(
+                x=audio_signal,
+                att_mask=att_mask,
+                pos_emb=pos_emb,
+                pad_mask=pad_mask,
+                cache_last_channel=cache_last_channel,
+                cache_last_time=cache_last_time,
+                cache_last_channel_next=cache_last_channel_next,
+                cache_last_time_next=cache_last_time_next,
+            )
+
+            if self.reduction_position == lth:
+                audio_signal, length = self.reduction_subsampling(x=audio_signal, lengths=length)
+                max_audio_length = audio_signal.size(1)
+                # Don't update the audio_signal here because then it will again scale the audio_signal
+                # and cause an increase in the WER
+                _, _, pad_mask, att_mask, pos_emb, cache_last_channel_next = self._create_masks(
+                    audio_signal, max_audio_length, length, cache_last_channel=cache_last_channel_next
+                )
+
+        if self.out_proj is not None:
+            audio_signal = self.out_proj(audio_signal)
+
+        # Reduction
+        if self.reduction_position == -1:
+            audio_signal, length = self.reduction_subsampling(x=audio_signal, lengths=length)
+
+        audio_signal = torch.transpose(audio_signal, 1, 2)
+
+        if cache_last_channel is not None:
+            return audio_signal, length, cache_last_channel_next, cache_last_time_next
+        else:
+            return audio_signal, length
+
+    def _create_masks(self, audio_signal, max_audio_length, length, cache_last_channel=None):
```

Comment:
First part of the code related to caching should not be inside create_masks. It should just creates the maskings.

Reply:
Pushed the code with caching outside create_masks